### DISCUSSION TITLE
CM4 USB power

### DISCUSSION THREAD

**Discussion by rpisnd** [rsta2/circle#333] [Q&A] [answered]
I am experimenting with Circle on a CM4 / CM4 IO Board (fairly new to CM4).
Somehow I cannot get the USB ports powered upon boot.
The config.txt I am using is

armstub=armstub7-rpi4.bin
max_framebuffers=2

dtoverlay=dwc2,dr_mode=host
otg_mode=1

Thank you for any tips / recommendations.
t*no

**Comment by rsta2** [comment, marked as answer]
For USB operation on the CM4 you have to define the system option `USE_XHCI_INTERNAL` before build. Just add the following line to your *Config.mk* file:

```
DEFINE += -DUSE_XHCI_INTERNAL
```

This enables the support for the internal xHCI USB host controller of the CM4. The DWC2 OTG USB controller is not supported by Circle for the CM4.

In *config.txt* you need the following settings for USB in AArch64 mode:

```
arm_64bit=1
kernel=kernel8-rpi4.img
armstub=armstub7-rpi4.bin
otg_mode=1
```

Be sure to have the device tree blobs *bcm2711-rpi-4-b.dtb* and *bcm2711-rpi-cm4.dtb* on the SD card.

Finally the following setting in the file *cmdline.txt* helps, if some USB device is not detected:

```
usbpowerdelay=2000
```

**Reply by rpisnd** [reply]
Thank you very much - works perfectly!
